In [ ]:
n = int(input())

for i in range(n):
    ans = i**2
    print(ans)

3
0
1
4


In [ ]:
from array import array
n = int(input())
arr = list(map(int, input().split()))


unique_score = list(set(arr))
unique_score.sort(reverse=True)
print(arr[0])

5
2 3 6 6 5
2


In [ ]:
n = int(input("Enter the number of elements: "))
arr = list(map(int, input("Enter the elements separated by space: ").split()))

if len(arr) < n:
    print("Error: The number of elements in the list is less than the specified index.")
else:
    arr.sort()
    print(arr[n-2])

Enter the number of elements: 5
Enter the elements separated by space: 23665
Error: The number of elements in the list is less than the specified index.


In [ ]:
# Read input
n = int(input())
scores = list(map(int, input().split()))

# Find the unique scores and sort them in descending order
unique_scores = list(set(scores))
unique_scores.sort(reverse=True)

# The runner-up score is the second highest unique score
if len(unique_scores) < 2:
    print("No runner-up score available.")
else:
    print(unique_scores[1])


5
23665
No runner-up score available.


In [ ]:
# Step 1: Read the number of students
n = int(input("Enter the number of students: "))

# Step 2: Store each student's name and grade in a nested list
students = []
for _ in range(n):
    name = input()
    grade = float(input())
    students.append([name, grade])

# Step 3: Find the second lowest grade
grades = sorted(set(student[1] for student in students))
second_lowest_grade = grades[1]

# Step 4: Extract the names of students with the second lowest grade
second_lowest_students = [student[0] for student in students if student[1] == second_lowest_grade]

# Step 5: Sort the names alphabetically and print them
for name in sorted(second_lowest_students):
    print(name)



# 5
# # Harry
# # 37.21
# # Berry
# # 37.21
# # Tina
# # 37.2
# # Akriti
# # 41
# # Harsh
# # 39



In [ ]:
if __name__ == '__main__':
    n = int(input())
    student_marks = {}

    # Reading student data
    for _ in range(n):
        name, *line = input().split()
        scores = list(map(float, line))
        student_marks[name] = scores

    query_name = input()

    # Calculating the average marks for the queried student
    query_scores = student_marks[query_name]
    average_marks = sum(query_scores) / len(query_scores)

    # Printing the average marks formatted to 2 decimal places
    print(f"{average_marks:.2f}")



# 3
# Krishna 67 68 69
# Arjun 70 98 63
# Malika 52 56 60
# Malika


In [ ]:
if __name__ == '__main__':
    n = int(input())
    integer_list = tuple(map(int, input().split()))

    ans = hash(integer_list)
    print(ans)

2
1 2
-3550055125485641917


In [ ]:
strng = input()
my_dict={}
strng = strng.lower()
for word in strng:
    word.count(word)
    my_dict[word]==my_dict.get(word,0)+1
print(my_dict)

AlmaBetter


KeyError: 'a'

In [ ]:
def is_prime(n):
    count = 0
    i = 1
    while(i<=n):
        if n%i==0:
            count+=1
        i = i+1
    if count==2:
        return True
    else:
        return False


def prime_importance(sentence):
    words = sentence.split()
    prime_words = [word for word in words if is_prime(len(word))]

    return " ".join (prime_words)

sentence = input()
print(prime_importance(sentence))

The quick brown fox jumps over the lazy dog.
The quick brown fox jumps the


# Write a python program to demonstrare different number data types.

In [ ]:
a = 34
b = 8.9
c = 34j

print(type(a))
print(type(b))
print(type(c))

<class 'int'>
<class 'float'>
<class 'complex'>


# Write a program to demonstrate different arithmetic operations on numbers in python .

In [ ]:
a = 45+65
b = 45-65
c = 45*65
d = 45/65
e = 45//65
f = 45%65

print(a)
print(b)
print(c)
print(d)
print(e)
print(f)

110
-20
2925
0.6923076923076923
0
45


# Write a apython program to create, concatenate and accessing the substring and also print the string

In [ ]:
fname= 'Siddharh'

lname = 'Deshpande'

full_name = fname + ' ' + lname

print(full_name)

sub_string = full_name[0:3]
print(sub_string)

Siddharh Deshpande
Sid


In [ ]:
n = int(input())

for i in range(n):
  for j in range(i):
    print("* ",end ="")

  print(" ")

for i in range(n,0,-1):
  for j in range(i):
    print("* ",end ="")
  print(" ")

5
 
*  
* *  
* * *  
* * * *  
* * * * *  
* * * *  
* * *  
* *  
*  


# Prime numbers between specified range

In [ ]:
lower_range = int(input("Enter the lower range: "))
upper_range = int(input("Enter the upper range: "))

for num in range(lower_range, upper_range + 1):
    if num > 1:
        for i in range(2, num):
            if (num % i) == 0:
                break
        else:
            print(num)

Enter the lower range: 10
Enter the upper range: 100
11
13
17
19
23
29
31
37
41
43
47
53
59
61
67
71
73
79
83
89
97


# given an un sorted integers list of numbers,
# return the first missing positive integers
# i/p [1,3,-5,0,2,-1]
# o/p 4

In [ ]:
def first_missing_positive(nums):
    n = len(nums)

    for i in range(n):
        while 1 <= nums[i] <= n and nums[nums[i] - 1] != nums[i]:
            nums[nums[i] - 1], nums[i] = nums[i], nums[nums[i] - 1]

    for i in range(n):
        if nums[i] != i + 1:
            return i + 1

    return n + 1

# Example usage
nums = [1, 3, -5, 0, 2, -1]
print(first_missing_positive(nums))

4


In [ ]:
strng = 'The quick brown fox jumps over the lazy dog'
final = strng.count('the')
print(final)

1


In [ ]:
strng = 'The quick brown fox jumps over the lazy dog'
final = strng.lower().count('the')
print(final)

2


In [ ]:
strng = 'The quick brown fox jumps over the lazy dog'
strng = strng.lower()  # Convert to lowercase to match case insensitively

words = strng.split()  # Split the string into words

count = 0
for word in words:
    if word == 'the':
        count += 1

print(count)


2


In [ ]:
import pandas as pd
import numpy as np

# Creating the DataFrame
data = {
    'a': [1, 2, 2, 2, 3, 3, 3, 4],
    'b': [5.0, 6.0, np.nan, 8.0, 3.0, 3.0, np.nan, 4.0]
}
df = pd.DataFrame(data)
df

a    b
0  1  5.0
1  2  6.0
2  2  NaN
3  2  8.0
4  3  3.0
5  3  3.0
6  3  NaN
7  4  4.0

In [ ]:

# Function to fill NaN values with the mean of corresponding 'a' values
def fill_nan_with_mean(df):
    # Iterate over each unique value in column 'a'
    for value in df['a'].unique():
        # Calculate the mean of 'b' for the current 'a' value, ignoring NaNs
        mean_value = df.loc[df['a'] == value, 'b'].mean()
        # Fill NaNs in 'b' where 'a' is the current value with the calculated mean
        df.loc[(df['a'] == value) & (df['b'].isna()), 'b'] = mean_value
    return df

# Applying the function to the DataFrame
df = fill_nan_with_mean(df)

# Displaying the updated DataFrame
print(df)

   a    b
0  1  5.0
1  2  6.0
2  2  7.0
3  2  8.0
4  3  3.0
5  3  3.0
6  3  3.0
7  4  4.0


In [ ]:
# Group by 'a' and transform 'b' to fill NaN values with the mean of the group
df['b'] = df.groupby('a')['b'].transform(lambda x: x.fillna(x.mean()))

# Display the updated DataFrame
print(df)

   a    b
0  1  5.0
1  2  6.0
2  2  7.0
3  2  8.0
4  3  3.0
5  3  3.0
6  3  3.0
7  4  4.0


In [ ]:
my_dict = {'a':1,'b':2,'c':1}
my_dict

{'a': 1, 'b': 2, 'c': 1}

In [ ]:
new_dict = {}
for key,value in my_dict.items():
    if value not in new_dict:
        new_dict[value] = [key]
    else:
        new_dict[value].append(key)

In [ ]:
new_dict

{1: ['a', 'c'], 2: ['b']}

In [ ]:
print(min(max(False, -3, -4), 2,7))

False


1. Replace all occurrences of 'xy' with 'yyx' untill there is not a single occurrence of 'xy' throughout the string.
2. Create a numpy array of 5 elements for expenses for 5 years. Create another numpy array to store 5 revenues for 5 years. Calculate and store the profit (revenue - expense) in the final numpy array and display the array.
3. Write a python program to print all Armstrong numbers within a given inputted range. See Google to know what is an Armstrong number.
4. Write a python program to print the substring of length 2 characters, which has the highest number of occurring frequency in the string.
5. Input a string from user. Input a substring from user. Find how many times the substring is present in the string.

SQL :
1. Inner, left, right, outer joins.
2. Date and time functions.
3. Round function and float formatting.
4. Aggregate function and window functions.
5. Imp : Group by clause, like clause, regexp clause.
6. Understand queries with aggregate function and group by in same query.

In [ ]:
#1
strng = input()
while 'xy' in strng:
    strng = strng.replace('xy','yyx')
    print(strng)

xyyxyxxy
yyxyyyxxyyx
yyyyxyyxyyxyx
yyyyyyxyyyxyyyxx
yyyyyyyyxyyyyxyyxx
yyyyyyyyyyxyyyyyxyxx
yyyyyyyyyyyyxyyyyyyxxx
yyyyyyyyyyyyyyxyyyyyxxx
yyyyyyyyyyyyyyyyxyyyyxxx
yyyyyyyyyyyyyyyyyyxyyyxxx
yyyyyyyyyyyyyyyyyyyyxyyxxx
yyyyyyyyyyyyyyyyyyyyyyxyxxx
yyyyyyyyyyyyyyyyyyyyyyyyxxxx


In [ ]:
#2.
import numpy as np
expenses = np.array([5678,6000,10000,8888,9819])
revenue = np.array([10000,15000,20000,18000,12222])
final = revenue-expenses
print(final)

[ 4322  9000 10000  9112  2403]


In [ ]:
#3

def is_armstrong(num):
    sum = 0
    temp = num
    while temp > 0:
        digit = temp % 10
        sum += digit ** 3
        temp //= 10
    return sum == num

a = int(input())
b = int(input())

for num in range(a, b + 1):
    if is_armstrong(num):
        print(num)

200
600
370
371
407


In [ ]:
#4
def most_frequent_substring(s):
    # Generate all possible substrings of length 2
    n = len(s)
    max_count = 0
    most_frequent = ''

    # Iterate through each possible substring of length 2
    for i in range(n - 1):
        current_substring = s[i:i+2]
        count = 0

        # Count occurrences of the current substring
        for j in range(n - 1):
            if s[j:j+2] == current_substring:
                count += 1

        # Update most frequent substring if necessary
        if count > max_count:
            max_count = count
            most_frequent = current_substring

    return most_frequent, max_count

# Example usage
string = "ababcbabc"
result = most_frequent_substring(string)
print(f"The substring of length 2 with the highest frequency is '{result[0]}' with a frequency of {result[1]}")


The substring of length 2 with the highest frequency is 'ab' with a frequency of 3


In [ ]:
#5
string = input()
substring = input()
count = string.count(substring)
print(count)

abcbacabcbbbab
ab
3


In [ ]:
def remove_duplicates(s):
    seen = set()
    result = []

    for char in s:
        if char not in seen:
            seen.add(char)
            result.append(char)

    return ''.join(result)

# Example usage
input_string = input("Enter a string: ")
output_string = remove_duplicates(input_string)
print("String after removing duplicates:", output_string)

Enter a string: siddharth deshpande
String after removing duplicates: sidhart epn


# Write a python progrm to return common letters bewtwwn 2 strings

In [ ]:
strng1 = input()
strng2 = input()

strng1 = strng1.lower()
strng2 = strng2.lower()

new_str = set(strng1)
new_str2 = set(strng2)

final_output = new_str.intersection(new_str2)
print(final_output)

Parth
Siddharth
{'r', 't', 'a', 'h'}


# count frequency of words in a string

In [ ]:
strng = input()
li = strng.split()
my_dict = {}
for word in li:
  if word not in my_dict.keys():
    my_dict[word] = 0
  my_dict[word] += 1
print(my_dict)

Sheena loves apple and mango but her sister loves guava and apple
{'Sheena': 1, 'loves': 2, 'apple': 2, 'and': 2, 'mango': 1, 'but': 1, 'her': 1, 'sister': 1, 'guava': 1}


# convert 2 list into dictionary]

In [ ]:
keys = eval(input())
values = eval(input())
result = dict(zip(keys,values))
print(result)

['sid','ukku','parth']
[124,567,999]
{'sid': 124, 'ukku': 567, 'parth': 999}


# Find missing number from array

In [ ]:
import numpy as np
arr = np.array([1,2,4,5,6,7,8,9,10])
n = arr[-1]
sum1 = 0
total = n*(n+1)/2
for i in arr:
  sum1 = sum1+i
print(total-sum1)

3.0


# Find out the pairs with given sum values of an array

In [ ]:
import numpy as np
def two_sum(arr,sum):
  arr.sort()
  left = 0
  right = len(arr)-1
  while(left<=right):
    if arr[left] + arr[right] < sum:
      left+=1
    elif arr[left] + arr[right] > sum:
      right-=1
    else:
      print(arr[left],arr[right])
      left+=1
      right-=1

arr = np.array([8,9,5,0,1,16,14,3])
sum = 17
print(two_sum(arr,sum))

1 16
3 14
8 9
None


# Minimum difference element from the array

# Same as the code for maximum difference from the array
# but here we have to use > sign and instead of +infinite we use -infinite


In [ ]:
def min_diff_element(arr):
  arr.sort()
  n = len(arr)-1
  min_diff = 9999*999
  for i in range(n):
    if arr[i+1]-arr[i]< min_diff:
      min_diff = arr[i+1] - arr[i]
    return min_diff

arr = [1,3,15,11,2]
print(min_diff_element(arr))


1


# Rotated String

In [ ]:
def rotated_strng(name):
  n = len(name)
  temp = name + name
  for i in range(n):
    for j in range(n):
      print(temp[i+j],end = "")
    print()

name = 'sid'
rotated_strng(name)

sid
ids
dsi


# Reverse words in string

In [ ]:
def rev_words_strng(strng):
    if len(strng) == 1:
        return strng

    strng1 = strng.split(" ")
    rev_all = ""

    for word in strng1:
        rev = ''.join(reversed(word))
        rev_all += rev + " "

    return rev_all.strip()

strng = 'sid deshpande'
print(rev_words_strng(strng))


dis ednaphsed


# First non repeating character

In [ ]:
def first_non_repeating_char(strng):
    my_dict = {}
    size = len(strng)
    for i in range(size):
      key = strng[i]
      if key not in my_dict.keys():
        my_dict[key] = 1
      else:
        my_dict[key] += 1
    print(my_dict)


    for key,value in my_dict.items():
      if value == 1:
        print(key)
        break


strng = 'siddeshpandesdes'
first_non_repeating_char(strng)

{'s': 4, 'i': 1, 'd': 4, 'e': 3, 'h': 1, 'p': 1, 'a': 1, 'n': 1}
i


In [ ]:
a = [1,2,3,4,5,6]

for a[-1] in a:
  print(a[-1])

1
2
3
4
5
5


In [ ]:
a = [1,2,3,4,5,6]

for a[-1] in a:
  print(a)

[1, 2, 3, 4, 5, 1]
[1, 2, 3, 4, 5, 2]
[1, 2, 3, 4, 5, 3]
[1, 2, 3, 4, 5, 4]
[1, 2, 3, 4, 5, 5]
[1, 2, 3, 4, 5, 5]


STING SWAP

In [ ]:
# i/p Python Exercise
# o/p pYTHON eXERCISE

strng = input()
res_str = ""
for item in strng:
  if item.isupper():
    res_str +=item.lower()
  else:
    res_str +=item.upper()
print(res_str)

Python Exercise
pYTHON eXERCISE


In [ ]:
lst = [1,2,3,0,4,5,0]
result = []
for num in lst:
  if num!=0:
    result.append(num)

for num in lst:
  if num == 0:
    result.append(num)
print(result)

# Convert the list to a string
result_string = ','.join(map(str, result))

print(result_string)

[1, 2, 3, 4, 5, 0, 0]
1,2,3,4,5,0,0


In [ ]:
lst = ["𝐌𝐨𝐧, 𝐓𝐮𝐞, 𝐖𝐞𝐝, 𝐓𝐡𝐮, 𝐅𝐫𝐢, 𝐒𝐚𝐭, 𝐒𝐮𝐧"]
lst = lst[0].split(',')
print(lst)
print(lst[::5])

['𝐌𝐨𝐧', ' 𝐓𝐮𝐞', ' 𝐖𝐞𝐝', ' 𝐓𝐡𝐮', ' 𝐅𝐫𝐢', ' 𝐒𝐚𝐭', ' 𝐒𝐮𝐧']
['𝐌𝐨𝐧', ' 𝐒𝐚𝐭']


# Write a Python program to find the frequency of a given item(string or int) in a tuple.

Sample Output

(2, 4, 5, 6, 2, 3, 4, 4, 7)

4

Output- 3

Example 1:
Input:
w,3,r,e,s,o,u,r,c,e,s s

Output:
2

In [ ]:
integer = eval(input())
a = int(input())
count = 0

for num in integer:
  if a ==  num:
    count+=1
print(count)

(1,2,1,1,1,1,1,1,1,2,5,6,7,8,9)
1
8


In [ ]:
s = input()
ch = input()

if s[0]==ch:
  print("True")
else:
  print("False")

python
p
True


In [ ]:
s = input()
ch = input()

if s.startswith(ch):
  print("True")
else:
  print("False")

Python
P
True


Armstrong number

In [ ]:
n = int(input())

sum = 0
temp = n

while temp>0:
  digit = temp%10
  sum = sum+digit**3
  temp = temp//10

if sum == n:
  print("Armstrong")
else:
  print("Not Armstrong")


153
Armstrong


Harmonic sum

In [ ]:
n = int(input())
sum = 0
for i in range(1,n+1):
  sum = sum + 1/i
print(sum)

7
2.5928571428571425


# Mingle to single

# def digital_root(n):
    if n == 0:
        return 0
    return 1 + (n - 1) % 9

    This is also a great logic of getting the sum


In [ ]:
n = int(input())
sum = 0
temp = n

while temp>0:
  digit = temp%10
  sum+=digit
  temp//=10
print("Step 1 Sum:",sum)

while sum > 9 :
  temp = sum
  sum = 0
  while temp>0:
    digit = temp%10
    sum+=digit
    temp//=10
print("Step 2 Sum:",sum)


88
Step 1 Sum: 16
Step 2 Sum: 7


# Words of prime importance

In [ ]:
def is_prime(n):
  count = 0
  i = 1
  while i<=n:
    if n%i == 0:
      count+=1
    i+=1
  if count == 2:
    return True
  else:
    return False

def importance_of_words(s):
  words = s.split()
  prime_word = [word for word in words if is_prime(len(word))]
  return " ".join(prime_word)

s = input()
print(importance_of_words(s))

I am Siddharth Deshpande Father of vedant gupta 
am of gupta


# Closest Palindrome

In [ ]:
s = input()
mid = len(s)//2
r = s[:mid]
r+=s[mid::-1]
print(r)

madan
madam


# is_pangram

In [ ]:
strng = input()
strng = strng.lower().split()
strng = ''.join(strng)

unique_char  = set(strng)
is_pangram = len(unique_char) == 26 and all(char in unique_char for char in 'abcdefghijklmnopqrstuvwxyz')
print(is_pangram)

The quick brown fox jumps over a lazy dog 
True


# Return the indices of list

In [ ]:
lst = eval(input())
n = int(input())
lst1 = []

for i in lst:
    if i<n:
        lst1.append(lst.index(i))
print(lst1)

[45,78,65,100,1000,2000]
2000
[0, 1, 2, 3, 4]


In [ ]:
lst = eval(input())
res = [i for i in lst if i is not None]
print(res)

[12, 0, None, 23, None, -55, 234, 89, None, 0, 6, -12]
[]


In [ ]:
lst = eval(input())
lst
res = [x.index(lst) for x in lst if x is None ]
print(res)

[[1, None, 5, 4,None, 0, None, None]]
[]


In [ ]:
lst = eval(input())  # Input should be a list, e.g., [1, None, 5, 4, None, 0, None, None]
indices = []

for index, value in enumerate(lst):
    if value is None:
        indices.append(index)

print(indices)

[None,4,None]
[0, 2]


In [ ]:
word = eval(input())
remove_word = eval(input())
result = [word for word in word if word not in remove_word]
print(result)
# ['red', 'green', 'blue', 'white', 'black', 'orange']
# ['green', 'orange']

['red', 'green', 'blue', 'white', 'black', 'orange']
['green', 'orange'] 
['red', 'blue', 'white', 'black']


# Count the Characters

# AlmaBetter
# o/p should be {a:2,b:1......}

In [ ]:
strng = input()
my_dict = {}
strng = strng.lower()
for char in strng:
  if char in my_dict:
    my_dict[char]+=1
  else:
    my_dict[char] = 1
print(my_dict)

Almabetter
{'a': 2, 'l': 1, 'm': 1, 'b': 1, 'e': 2, 't': 2, 'r': 1}


In [ ]:
def digital_root(n):
    if n == 0:
        return 0
    return 1 + (n - 1) % 9

# Example usage
n = int(input("Enter a positive integer: "))
if n <= 0:
    print("The number must be a positive integer.")
else:
    result = digital_root(n)
    print("The single-digit result is:", result)

Enter a positive integer: 987654
The single-digit result is: 3


# Print Space between characters

In [ ]:
n = eval(input())
output = []

for char,count in n.items():
  output.extend([char]*count)
print(" ".join(output))

{"f":1,"o":2}
f o o


# Strange Sort list

i/p-[1,3,4,5,11]

o/p-[1,11,3,5,4]

first smallest then largest and carry forward this

In [ ]:
# Safely parse the input
numbers = eval(input())
result = []

while numbers:
    # Find and remove the minimum element
    minimum = min(numbers)
    result.append(minimum)
    numbers.remove(minimum)

    if numbers:
        # Find and remove the maximum element
        maximum = max(numbers)
        result.append(maximum)
        numbers.remove(maximum)

print(result)


[2,8,7,15,6]
[2, 15, 6, 8, 7]


# Pascal Triangle

In [ ]:
from math import factorial
rows = int(input())

for n in range(rows):
  for space in range(1,rows-n):
    print(end=" ")
  for r in range(n+1):
    ncr = factorial(n)/(factorial(r)*factorial(n-r))
    print(ncr,end=" ")
  print(" ")

4
   1.0  
  1.0 1.0  
 1.0 2.0 1.0  
1.0 3.0 3.0 1.0  


# Logic of Palindrome

In [ ]:
n = int(input())

temp = n
rev = 0

while(temp>0):
  digit = temp%10
  rev = rev*10 + digit
  temp//=10

if rev == n:
  print("Palindrome")
else:
  print("Not Palindrome")

1331
Palindrome


# Logic Of Palindrome in string format

In [ ]:
strng = input()
strng = strng.lower()
if strng == strng[::-1]:
  print("Palindrome")
else:
  print("Not Palindrome")

Nayan
Palindrome


# Get next_date

In [ ]:
def is_leap_year(year):
  if year%4==0:
    if year%100==0:
      if year%400==0:
        return True
      else:
        return False
    else:
      return True
  else:
    return False

def is_days_in_month(month,year):
  if month in (1,3,5,7,8,10,12):
    return 31
  elif month in (4,6,9,11):
    return 30
  elif month == 2:
    if is_leap_year(year):
      return 29
    else:
      return 28

def get_next_date(year,month,day):
  next_day=day+1
  next_month = month
  next_year = year

  if next_day > is_days_in_month(next_month,next_year):
    next_day = 1
    next_month+=1

  if next_month>12:
    next_month = 1
    next_year+=1

  return next_year,next_month,next_day


year = int(input())
month = int(input())
day = int(input())

next_date = get_next_date(year,month,day)
print(f"The next date is [yyyy-mm-dd] {next_date}.")

2024
10
10
The next date is [yyyy-mm-dd] (2024, 10, 11).


In [ ]:
sum = 0
count = 0
number = 1

while number!=0:
    number = int(input())
    sum = sum+number
    count+=1

if count==0:
    print("Input some number")
else:
    print("Average and Sum of the above numbers are:",sum/(count-1),sum )

15
45
9
99
100
0
Average and Sum of the above numbers are: 53.6 268


In [ ]:
lst = eval(input())
lst1 = []
sum = 0
for i in lst:
  sum+=i
  lst1.append(sum)
print(lst1)

[10, 20, 30, 40, 50, 60, 7]
[10, 30, 60, 100, 150, 210, 217]


In [ ]:
list1 = eval(input())
list2= eval(input())

list1 = set(list1)
list2 = set(list2)

ans = list1.difference(list2)
ans = list(ans)
print(ans)

[1,2,3,4,5,6,7]
[4,5,6,7]
[1, 2, 3]


# Logic of LCM

In [ ]:
a = int(input())
b = int(input())

MaxNum = max(a,b)

while True:
  if MaxNum%a==0 and MaxNum%b==0:
    break
  MaxNum+=1
print(MaxNum)

6
8
24


# Logic of HCF

In [ ]:
a = int(input())
b = int(input())

MinNum = min(a,b)
HCF = 1

for i in range(1,MinNum+1):
  if a%i==0 and b%i==0:
    HCF = i
print(HCF)

10
20
10


# Number to english words

In [ ]:
digit_to_word = {
    '0':'Zero',
    '1':'One',
    '2':'Two',
    '3':'Three',
    '4':'Four',
    '5':'Five',
    '6':'Six',
    '7':'Seven',
    '8':'Eight',
    '9':'Nine',
}

def singe_digit_to_words(nums):
  single_digit = []

  for num in nums:
    if 0<num<10:
      single_digit.append(digit_to_word[str(num)])
  return single_digit

nums = eval(input())
print(singe_digit_to_words(nums))

[1,3,5,7,9]
['One', 'Three', 'Five', 'Seven', 'Nine']


# Count digit and letter in string

In [ ]:
s = input()

count_digit = 0
count_char = 0

for char in s:
  if char.isdigit():
    count_digit+=1
  else:
    count_char+=1

print(count_digit)
print(count_char)

"1234getonthedancefloor"
4
20


In [ ]:
lst = eval(input("Enter a list of lists: "))
k = int(input("Enter an index: "))

if not lst or k < 0 or k >= len(lst[0]):
    print("Invalid")
else:
    answer = [row[k] for row in lst if len(row) > k]
    print(answer)

TypeError: object of type 'int' has no len()

# Adding the string

In [ ]:
def concatenate_strng(lst,n):
  result = []
  for i in lst:
    result.append(n+str(i))
  return result

lst = eval(input())
n = input()
print(concatenate_strng(lst,n))

[1,2,3,4]
emp
['emp1', 'emp2', 'emp3', 'emp4']


# Star Patteern logic

In [ ]:
n = int(input())

for i in range(0,n):
  for j in range(i):
    print("#",end="")
  print(" ")

6
 
# 
## 
### 
#### 
##### 


In [ ]:
n = int(input())

for i in range(0,n):
  for space in range(0,n-i):
    print(" ",end = "")
  for j in range(i):
    print("&",end = "")
  print(" ")


6
       
     & 
    && 
   &&& 
  &&&& 
 &&&&& 


In [ ]:
n = int(input())

for i in range(0,n):
  for space in range(0,n-i):
    print(" ",end = "")
  for j in range(0,(2*i)-1):
    print("*",end = "")
  print(" ")

7
        
      * 
     *** 
    ***** 
   ******* 
  ********* 
 *********** 


# Fibonacci

In [ ]:
def fibonacci(num):
  x,y = 0,1
  for i in range(num):
    print(x,end=" ")
    x,y = y,x+y

num = int(input())
fibonacci(num)

10
0 1 1 2 3 5 8 13 21 34 

In [ ]:
n = int(input())

num = 1

for i in range(0,n):
  for j in range(0,i):
    print(num, end=" ")
    num+=1
  print(" ")

6
 
1  
2 3  
4 5 6  
7 8 9 10  
11 12 13 14 15  


# Given 2 integers a and b, task is to determine the sum of cubes of all numbers in the range from a to b

In [ ]:
a = int(input())
b = int(input())

sum = 0

for i in range(a,b+1):
  sum = sum+i**3
print(sum)


4
9
1989


# Given an integer N, Task is to calculate the sum of first 10 Multiples of N and print the result

In [ ]:
n = int(input())
sum = 0
for i in range(1,11):
  mult = i*n
  sum+=mult
print(sum)


10
550


# You are given a grocery list of items, where each item is represented by its name,price per unit,and quantity sold. Your task is to:

  1. Identify the item with highest total selling price.
  2. Calculate the total selling price of all items.
  3. Calculate average selling price of all items.

In [ ]:
lst = [["apple", 1, 5], ["orange", 10, 5], ["apple", 10, 5]]

# Identify the item with the highest total selling price
ans = max(lst, key=lambda x: x[1] * x[2])

# Calculate the total selling price of all items
sum = 0
for item in lst:
  sum+=item[1]*item[2]
ans1 = sum

# Calculate the average selling price of all items
ans2 = ans1 / len(lst)

# Print the results
print(f"Item with the highest total selling price: {ans}")
print(f"Total selling price of all items: {ans1}")
print(f"Average selling price of all items: {ans2}")


Item with the highest total selling price: ['orange', 10, 5]
Total selling price of all items: 105
Average selling price of all items: 35.0


In [ ]:
import numpy as np

arr = np.array([7,10,4,3,20,15])
n = int(input())

ans = np.sort(arr)
print(ans[n-1])

3
7


# If input is string then output should be rtsng

In [ ]:
def process_string(input_str):
    result = []
    temp = []

    for char in input_str:
        if char == 'i':
            # Reverse the collected characters and add them to the result
            result.extend(reversed(temp))
            # Clear the temporary list
            temp = []
        else:
            # Add characters to the temporary list
            temp.append(char)

    # Append any remaining characters after the last 'i'
    result.extend(temp)

    # Convert the list back to a string
    return ''.join(result)

# Example usage:
input_str = input()
output_str = process_string(input_str)
print(output_str)  # Output will be "rtsng"


print
rpnt


In [ ]:
def reverse_before_i(s):
    if 'i' in s:
        index_i = s.index('i')
        return s[:index_i][::-1] + s[index_i + 1:]
    return s

input_string = input()
output_string = reverse_before_i(input_string)
print(output_string)  # Output: rtsng

Vidhi
Vdhi


# Input :

8  – Value of N

[4,5,0,1,9,0,5,0] – Element of arr[O] to arr[N-1],While input each element is separated by newline

Output:

4 5 1 9 5 0 0 0

In [ ]:
lst = eval(input())

lst1 = []
result = []

for i in lst:
  if i==0:
    lst1.append(0)
  else:
    result.append(i)

result.extend(lst1)
print(result)

[4,5,0,1,9,0,5,0]
[4, 5, 1, 9, 5, 0, 0, 0]


# Joseph is learning digital logic subject which will be for his next semester. He usually tries to solve unit assignment problems before the lecture. Today he got one tricky question. The problem statement is “A positive integer has been given as an input. Convert decimal value to binary representation. Toggle all bits of it after the most significant bit including the most significant bit. Print the positive integer value after toggling all bits”.

Constrains-

1<=N<=100

Example 1:

Input :

10  -> Integer

Output :

5    -> result- Integer

Explanation:

Binary representation of 10 is 1010. After toggling the bits(1010), will get 0101 which represents “5”. Hence output will print “5”.

In [ ]:
def toggle_bits(n):
    # Convert the number to its binary representation (excluding the '0b' prefix)
    binary_rep = bin(n)[2:]

    # Toggle all bits in the binary representation
    toggled_binary = ''.join('1' if bit == '0' else '0' for bit in binary_rep)

    # Convert the toggled binary string back to a decimal number
    toggled_number = int(toggled_binary, 2)

    return toggled_number

# Input
n = int(input("Enter a positive integer: "))

# Output
print(toggle_bits(n))

Enter a positive integer: 8
7


# Given an integer array Arr of size N the task is to find the count of elements whose value is greater than all of its prior elements.

Note : 1st element of the array should be considered in the count of the result.

For example,

Arr[]={7,4,8,2,9}

As 7 is the first element, it will consider in the result.

8 and 9 are also the elements that are greater than all of its previous elements.

Since total of  3 elements is present in the array that meets the condition.

Hence the output = 3.

In [ ]:
lst = eval(input())

count = 1
for i in range(1,len(lst)):
  if lst[i]>lst[i-1]:
    count+=1
print(count)

[3,4,5,8,9]
5


# A supermarket maintains a pricing format for all its products. A value N is printed on each product. When the scanner reads the value N on the item, the product of all the digits in the value N is the price of the item. The task here is to design the software such that given the code of any item N the product (multiplication) of all the digits of value should be computed(price).

Example 1:

Input :

5244 -> Value of N

Output :
160 -> Price

Explanation:

From the input above

Product of the digits 5,2,4,4

5*2*4*4= 160

Hence, output is 160.

In [ ]:
n = int(input())
product = 1

while n>0:
  digit = n%10
  product*=digit
  n = n//10

print(product)

5244
160


# A furnishing company is manufacturing a new collection of curtains. The curtains are of two colors aqua(a) and black (b). The curtains color is represented as a string(str) consisting of a’s and b’s of length N. Then, they are packed (substring) into L number of curtains in each box. The box with the maximum number of ‘aqua’ (a) color curtains is labeled. The task here is to find the number of ‘aqua’ color curtains in the labeled box.

Note :

If ‘L’ is not a multiple of N, the remaining number of curtains should be considered as a substring too. In simple words, after dividing the curtains in sets of ‘L’, any curtains left will be another set(refer example 1)

Example 1:

Input :

bbbaaababa -> Value of str

3    -> Value of L

Output:

3   -> Maximum number of a’s

Explanation:

From the input given above.

Dividing the string into sets of 3 characters each

Set 1: {b,b,b}

Set 2: {a,a,a}

Set 3: {b,a,b}

Set 4: {a} -> leftover characters also as taken as another set

Among all the sets, Set 2 has more number of a’s. The number of a’s in set 2 is 3.

Hence, the output is 3.

In [ ]:
def max_aqua_curtains(s: str, L: int) -> int:
    max_a_count = 0
    N = len(s)

    # Traverse the string in chunks of size L
    for i in range(0, N, L):
        # Extract the substring
        substring = s[i:i+L]

        # Count the number of 'a's in the substring
        a_count = substring.count('a')

        # Update the maximum number of 'a's
        max_a_count = max(max_a_count, a_count)

    return max_a_count

# Example Usage
str_input = input()
L = int(input())
result = max_aqua_curtains(str_input, L)
print(result)


"ababababaabbabab"
4
2


# An international round table conference will be held in india. Presidents from all over the world representing their respective countries will be attending the conference. The task is to find the possible number of ways(P) to make the N members sit around the circular table such that.

The president and prime minister of India will always sit next to each other.

Example 1:

Input :

4   -> Value of N(No. of members)

Output :

12  -> Possible ways of seating the members

Explanation:

2  members should always be next to each other.

So, 2 members can be in 2!ways

Rest of the members can be arranged in (4-1)! ways.(1 is subtracted because the previously selected two members will be considered as single members now).

So total possible ways 4 members can be seated around the circular table 2*6= 12.

Hence, output is 12.

Example 2:

Input:

10  -> Value of N(No. of members)

Output :

725760 -> Possible ways of seating the members

Explanation:

2 members should always be next to each other.

So, 2 members can be in 2! ways

Rest of the members can be arranged in (10-1)! Ways. (1 is subtracted because the previously selected two members will be considered as a single member now).

So, total possible ways 10 members can be seated around a round table is

2*362880 = 725760 ways.

Hence, output is 725760.

In [ ]:
def factorial(n):
    if n == 0:
      return 1
    else:
      return n * factorial(n-1)

n = int(input())
result = factorial(n-1) * 2
print(result)

10
725760


# An intelligence agency has received reports about some threats. The reports consist of numbers in a mysterious method. There is a number “N” and another number “R”. Those numbers are studied thoroughly and it is concluded that all digits of the number ‘N’ are summed up and this action is performed ‘R’ number of times. The resultant is also a single digit that is yet to be deciphered. The task here is to find the single-digit sum of the given number ‘N’ by repeating the action ‘R’ number of times.

If the value of ‘R’ is 0, print the output as ‘0’.

Example 1:

Input :

99 -> Value of N

3  -> Value of R

Output :

9  -> Possible ways to fill the cistern.

Explanation:

Here, the number N=99

Sum of the digits N: 9+9 = 18
Repeat step 2 ‘R’ times i.e. 3 tims  (9+9)+(9+9)+(9+9) = 18+18+18 =54
Add digits of 54 as we need a single digit 5+4
Hence , the output is 9.

In [ ]:
n = int(input())
r = int(input())

sum = 0

while r > 0:
  while n > 0:
    digit = n % 10
    sum += digit
    n = n // 10
  r -= 1

  n = sum
  sum = 0

while n > 0:
  digit = n % 10
  sum += digit
  n = n // 10
print(sum)

1234
2
1


# Particulate matters are the biggest contributors to Delhi pollution. The main reason behind the increase in the concentration of PMs include vehicle emission by applying Odd Even concept for all types of vehicles. The vehicles with the odd last digit in the registration number will be allowed on roads on odd dates and those with even last digit will on even dates.

Given an integer array a[], contains the last digit of the registration number of N vehicles traveling on date D(a positive integer). The task is to calculate the total fine collected by the traffic police department from the vehicles violating the rules.

Note : For violating the rule, vehicles would be fined as X Rs.

Example 1:

Input :

4 -> Value of N

{5,2,3,7} -> a[], Elements a[0] to a[N-1], during input each element is separated by a new line

12 -> Value of D, i.e. date

200 -> Value of x i.e. fine

In [ ]:
n = int(input())            # Number of vehicles
d = int(input())            # Date
x = int(input())            # Fine amount
lst = eval(input())         # List of last digits of the registration numbers

count = 0

# Check each last digit in the list
for i in lst:
    if d % 2 == 0:          # Check if the date is even
        if i % 2 != 0:      # If date is even, fine vehicles with an odd last digit
            count += 1
    else:                   # If the date is odd
        if i % 2 == 0:      # Fine vehicles with an even last digit
            count += 1

# Calculate the total fine
ans = count * x

# Print the total fine
print(ans)

5
3
300
[2,5,1,6,8]
900


# Smallest element in the array

In [ ]:
import numpy as np

# lst = eval(input())

# print(min(lst))

# lst = eval(input())

# min = lst[0]

# for i in range(1,len(lst)):
#   if lst[i]<min:
#     min = lst[i]

# print(min)

lst = eval(input())

ans = sorted(lst)
print(ans[0])


1,2,3,4,0
0


# Second cmallest or second largest

In [ ]:
lst = eval(input())
ans = sorted(lst)
print(ans[1])

[1,2,3,4,5,6]
2


# The function accepts two positive integers ‘r’ and ‘unit’ and a positive integer array ‘arr’ of size ‘n’ as its argument ‘r’ represents the number of rats present in an area, ‘unit’ is the amount of food each rat consumes and each ith element of array ‘arr’ represents the amount of food present in ‘i+1’ house number, where 0 <= i

Note:

Return -1 if the array is null
Return 0 if the total amount of food from all houses is not sufficient for all the rats.
Computed values lie within the integer range.
Example:

Input:

r: 7
unit: 2
n: 8
arr: 2 8 3 5 7 4 1 2

In [ ]:
# Input the number of rats and the amount of food each rat consumes
r = int(input("Enter the number of rats: "))
unit = int(input("Enter the amount of food each rat consumes: "))
lst = eval(input("Enter the list of food quantities in each house: "))

# Check if the list is empty
if not lst:
    print(-1)  # Return -1 if the array is null
else:
    total_required_food = r * unit  # Total food required for all rats
    total_available_food = 0  # Initialize total available food to 0

    # Calculate total available food in all houses
    for food in lst:
        total_available_food += food

    # Check if the total available food is less than the required food
    if total_available_food < total_required_food:
        print(0)  # Not enough food for all rats
    else:
        accumulated_food = 0  # Initialize accumulated food to 0
        houses_needed = 0  # Initialize the count of houses needed

        # Loop through the list to count how many houses are needed
        for food in lst:
            accumulated_food += food
            houses_needed += 1
            if accumulated_food >= total_required_food:
                break  # Stop when enough food is collected

        print(houses_needed)  # Output the number of houses needed


Enter the number of rats: 7
Enter the amount of food each rat consumes: 2
Enter the list of food quantities in each house: [2,8,3,5,7,4,1,2]
4


# The function accepts an integer array ‘arr’, its length and two integer variables ‘num’ and ‘diff’. Implement this function to find and return the number of elements of ‘arr’ having an absolute difference of less than or equal to ‘diff’ with ‘num’.
Note: In case there is no element in ‘arr’ whose absolute difference with ‘num’ is less than or equal to ‘diff’, return -1.

Example:
Input:

arr: 12 3 14 56 77 13
num: 13
diff: 2
Output:
3

Explanation:
Elements of ‘arr’ having absolute difference of less than or equal to ‘diff’ i.e. 2 with ‘num’ i.e. 13 are 12, 13 and 14.

In [ ]:
lst  = eval(input())
num = int(input())
diff = int(input())

count = 0

for i in lst:
  if abs(i-num)<=diff:
    count+=1

if count==0:
  print(-1)
else:
  print(count)

[12,3,14,56,77,13]
13
2
3


# The function accepts two integers n, m as arguments Find the sum of all numbers in range from 1 to m(both inclusive) that are not divisible by n. Return difference between sum of integers not divisible by n with sum of numbers divisible by n.

Assumption:

n>0 and m>0
Sum lies between integral range
Example

Input
n:4
m:20
Output
90

Explanation

Sum of numbers divisible by 4 are 4 + 8 + 12 + 16 + 20 = 60
Sum of numbers not divisible by 4 are 1 +2 + 3 + 5 + 6 + 7 + 9 + 10 + 11 + 13 + 14 + 15 + 17 + 18 + 19 = 150
Difference 150 – 60 = 90

In [ ]:
n = int(input())
m = int(input())

sum = 0
sum1 = 0

for i in range(1,m+1):
  if i%n==0:
    sum+=i
  else:
    sum1+=i

print(sum1-sum)


3
10
19


# Method 2

In [ ]:
n = int(input("Enter n: "))
m = int(input("Enter m: "))

# Sum of all numbers from 1 to m
total_sum = m * (m + 1) // 2

# Largest integer p such that p * n <= m
p = m // n

# Sum of all multiples of n from 1 to m
sum_divisible_by_n = n * p * (p + 1) // 2

# Sum of numbers not divisible by n
sum_not_divisible_by_n = total_sum - sum_divisible_by_n

# Result: sum_not_divisible_by_n - sum_divisible_by_n
result = sum_not_divisible_by_n - sum_divisible_by_n

print(result)


Enter n: 3
Enter m: 10
19


# The function accepts an integers arr of size ’length’ as its arguments you are required to return the sum of second largest  element from the even positions and second smallest from the odd position of given ‘arr’

Assumption:

All array elements are unique
Treat the 0th position as even
NOTE

Return 0 if array is empty
Return 0, if array length is 3 or less than 3
Example

Input

arr:3 2 1 7 5 4

Output

7

Explanation

Second largest among even position elements(1 3 5) is 3
Second smallest among odd position element is 4
Thus output is 3+4 = 7

In [ ]:
lst = eval(input())
lst1 = []
lst2 = []

if not lst:
  print(0)

if len(lst)<=3:
  print(0)

for i in range(0,len(lst)):
  if i%2==0:
    lst1.append(lst[i])
  else:
    lst2.append(lst[i])

ans1 = sorted(lst1)
ans2 = sorted(lst2)

print((ans1[-2]+ans2[1]))

[3,2,1,7,5,4]
7


# The function accepts an integers sum and an integer array arr of size n. Implement the function to find the pair, (arr[j], arr[k]) where j!=k, Such that arr[j] and arr[k] are the least two elements of array (arr[j] + arr[k] <= sum) and return the product of element of this pair

NOTE

Return -1 if array is empty or if n<2
Return 0, if no such pairs found
All computed values lie within integer range
Example

Input

sum:9

size of Arr = 7

Arr:5 2 4 3 9 7 1

Output

2

Explanation

Pair of least two element is (2, 1) 2 + 1 = 3 < 9, Product of (2, 1) 2*1 = 2. Thus, output is 2

In [ ]:
a = int(input())
lst = eval(input())

if not lst or len(lst)<2:
  print(0)

ans = sorted(lst)

if ans[0]+ans[1]<=a:
  print(ans[0]*ans[1])
else:
  print(0)

4
[9,8,3,-7,3,9]
-21


# The function accept positive integer n and num Implement the function to calculate the n-base equivalent of num and return the same as a string

Steps:

Divide the decimal number by n,Treat the division as the integer division
Write the the remainder (in  n-base notation)
Divide the quotient again by n, Treat the division as integer division
Repeat step 2 and 3 until the quotient is 0
The n-base value is the sequence of the remainders from last to first
Assumption:
1 < n < = 36

Example

Input
n: 12
num: 718

Output
4BA

Explanation

num       Divisor           quotient              remainder
718           12               59                 10(A)

59             12               4                   11(B)

4               12                0                   4(4)

In [ ]:
n = int(input("Enter the base (n): "))
num = int(input("Enter the number (num): "))

char_output = ""

while num != 0:
    remainder = num % n
    quotient = num // n

    # Map remainder to character (A=10, B=11, ..., Z=35)
    if remainder >= 10:
        char_output = chr(55 + remainder) + char_output
    else:
        char_output = chr(48 + remainder) + char_output  # Numbers 0-9

    num = quotient

print(char_output)





Enter the base (n): 21
Enter the number (num): 5678
CI8


# The function accepts a string “str” of length ‘n’, that contains alphabets and hyphens (-). Implement the function to move all hyphens(-) in the string to the front of the given string.

NOTE:- Return null if str is null.

Example :-

Input:
str.Move-Hyphens-to-Front
Output:
—MoveHyphenstoFront
Explanation:-

The string “Move-Hyphens -to-front” has 3 hyphens (-), which are moved to the front of the string, this output is “— MoveHyphen”

Sample Input

Str: String-Compare
Sample Output-

-StringCompare

In [ ]:
strng = input()

strng.split(" ")

for i in strng:
  if i == "-":
    strng = strng.replace(i,"")
    strng = "-"+strng

print(strng)

string-compare
-stringcompare


# he functions accepts two numbers ‘num1’ and ‘num2’ as its arguments. You are required to calculate and return  the total number of carries generated while adding digits of two numbers ‘num1’ and ‘ num2’.

Assumption: num1, num2>=0

Example:

Input
Num 1: 451
Num 2: 349
Output
2
Explanation:

Adding ‘num 1’ and ‘num 2’ right-to-left results in 2 carries since ( 1+9) is 10. 1 is carried and (5+4=1) is 10, again 1 is carried. Hence 2 is returned.

In [ ]:
num1 = int(input())
num2 = int(input())

carry = 0
count_carry = 0

while num1 > 0 or num2 > 0 or carry > 0:
    digit1 = num1 % 10
    digit2 = num2 % 10

    # Sum the digits plus any carry from the previous step
    sum_digits = digit1 + digit2 + carry

    # If the sum is 10 or greater, a carry occurs
    if sum_digits >= 10:
        carry = 1
        count_carry += 1
    else:
        carry = 0

    # Move to the next digit
    num1 //= 10
    num2 //= 10

print(count_carry)


23
564
0


# The function accepts a string  ‘ str’ of length n and two characters ‘ch1’ and ‘ch2’ as its arguments . Implement the function to modify and return the string ‘ str’ in such a way that all occurrences of ‘ch1’ in original string are replaced by ‘ch2’ and all occurrences of ‘ch2’  in original string are replaced by ‘ch1’.

Assumption: String Contains only lower-case alphabetical letters.

Note:

Return null if string is null.
If both characters are not present in string or both of them are same , then return the string unchanged.
Example:

Input:
Str: apples
ch1:a
ch2:p
Output:
paales
Explanation:

‘A’ in original string is replaced with ‘p’ and ‘p’ in original string is replaced with ‘a’, thus output is paales.

In [ ]:
def swap_characters(strng, ch1, ch2):
    if not strng:
        return None

    # If both characters are the same or not in the string, return the string unchanged
    if ch1 == ch2 or (ch1 not in strng and ch2 not in strng):
        return strng

    # Temporarily replace ch1 with a placeholder to avoid conflicts
    placeholder = '#'
    strng = strng.replace(ch1, placeholder)
    strng = strng.replace(ch2, ch1)
    strng = strng.replace(placeholder, ch2)

    return strng

# Example usage
strng = input("Enter the string: ")
ch1 = input("Enter the first character: ")
ch2 = input("Enter the second character: ")

result = swap_characters(strng, ch1, ch2)
print(result)


Enter the string: siddharth
Enter the first character: d
Enter the second character: s
dissharth


# The function accepts 3 positive integers ‘a’ , ‘b’ and ‘c ‘ as its arguments. Implement the function to return.

( a+ b ) , if c=1
( a – b ) , if c=2
( a * b ) ,  if c=3
(a / b) ,  if c =4
Assumption : All operations will result in integer output.

Example:

Input
c :1
a:12
b:16
Output:
Since ‘c’=1 , (12+16) is performed which is equal to 28 , hence 28 is returned.

In [ ]:
c = int(input())
a = int(input())
b = int(input())

if c==1:
  print(a+b)
elif c==2:
  print(a-b)
elif c==3:
  print(a*b)
elif c==4:
  print(a/b)
else:
  print("error")

2
6
10
-4


# You have to find and return the number between ‘a’ and ‘b’ ( range inclusive on both ends) which has the maximum exponent of 2.

The algorithm to find the number with maximum exponent of 2 between the given range is

Loop between ‘a’ and ‘b’. Let the looping variable be ‘i’.
Find the exponent (power) of 2 for each ‘i’ and store the number with maximum exponent of 2 so faqrd in a variable , let say ‘max’. Set ‘max’ to ‘i’ only if ‘i’ has more exponent of 2 than ‘max’.
Return ‘max’.
Assumption: a <b

Note: If two or more numbers in the range have the same exponents of  2 , return the small number.

Example

Input:
7
12
Output:
8

In [ ]:
# a = int(input())
# b - int(input())

# for i in range(a,b+1):
#   if 2**3 > i:
#     max = 2**3

# print(max)

a = int(input())
b = int(input())

max_exponent = 0
max_number = a

for i in range(a, b + 1):
    exponent = 0
    num = i

    while num % 2 == 0:
        num //= 2
        exponent += 1

    if exponent > max_exponent:
        max_exponent = exponent
        max_number = i
    elif exponent == max_exponent and i < max_number:
        max_number = i

print(max_number)

2
259
256


# The function accepts 2 positive integer ‘m’ and ‘n’ as its arguments.You are required to calculate the sum of numbers divisible both by 3 and 5, between ‘m’ and ‘n’ both inclusive and return the same.
Note
0 < m <= n

Example

Input:

m : 12

n : 50

Output

90

Explanation:
The numbers divisible by both 3 and 5, between 12 and 50 both inclusive are {15, 30, 45} and their sum is 90.
Sample Input
m : 100
n : 160
Sample Output
510

In [ ]:
n = int(input())
m = int(input())

sum = 0

for i in range(n,m+1):
  if i%3==0 and i%5==0:
    sum+=i
print(sum)

100
160
510


In [ ]:
n = int(input("Input the number for which u want to write the table:"))
sum = 0
for i in range(1,11):
  print(i*n)
  sum+=i*n
print(" ")
print("output sum =:",sum)

Input the number for which u want to write the table10
10
20
30
40
50
60
70
80
90
100
 
output sum =: 550


# Question: Write a program in C such that it takes a lower limit and upper limit as inputs and print all the intermediate palindrome numbers.

Test Cases:

TestCase 1:
Input :
10 , 80
Expected Result:
11 , 22 , 33 , 44 , 55 , 66 , 77.

Test Case 2:
Input:
100,200
Expected Result:
101 , 111 , 121 , 131 , 141 , 151 , 161 , 171 , 181 , 191

In [ ]:
n = int(input())
m = int(input())

for i in range(n,m+1):
  if str(i)==str(i)[::-1]:
    print(i,end = " ")

100
200
101 111 121 131 141 151 161 171 181 191 

# Q1) Given an array of integers where every element appears even number of times except one element which appears odd number of times, write a program to find that odd occurring element in O(log n) time. The equal elements must appear in pairs in the array but there cannot be more than two consecutive occurrences of an element.


5

2 2 3 1 1

Sample Output :

3

In [ ]:
lst = eval(input())

for i in lst:
  if lst.count(i)%2!=0:
    print(i)
    break

[2,2,3,1,1]
3


In [ ]:
lst = eval(input())  # Input list

result = 0
for i in lst:
    result ^= i

print(result)


[2,2,3,1,1]
3


In [ ]:
i = 2
while True:
  if i%3 == 0:
    break
  print(i)
  i += 2



2
4


In [ ]:
total={}
def insert(items):
  if items in total:
    total[items] += 1
  else:
    total[items] = 1
insert('Apple')
insert('Ball')
insert('Apple')
print (len(total))

2


In [ ]:
x = "abcdef"
i = "a"
while i in x:
  x = x[:-1]
  print(i, end = " ")

a a a a a a 

In [ ]:
def f1():
  x=100
  print(x)
  x=+1
f1()

100


In [ ]:
d = {0: 'a', 1: 'b', 2: 'c'}
for x, y in d.items():
  print(x, y)

0 a
1 b
2 c


In [ ]:
a={1:"A",2:"B",3:"C"}
print(a.get(1,4))

A


# Problem Statement – An automobile company manufactures both a two wheeler (TW) and a four wheeler (FW). A company manager wants to make the production of both types of vehicle according to the given data below:

1st data, Total number of vehicle (two-wheeler + four-wheeler)=v
2nd data, Total number of wheels = W
The task is to find how many two-wheelers as well as four-wheelers need to manufacture as per the given data.
Example :

Input :
200  -> Value of V
540   -> Value of W

Output :
TW =130 FW=70

Explanation:
130+70 = 200 vehicles
(70*4)+(130*2)= 540 wheels

In [ ]:
v = int(input())
w = int(input())

tw = 2*v - w//2
fw = w//2 - v

print("tw=",tw)
print("fw=",fw)

200
540
tw= 130
fw= 70


# Problem Statement – Given a string S(input consisting) of ‘*’ and ‘#’. The length of the string is variable. The task is to find the minimum number of ‘*’ or ‘#’ to make it a valid string. The string is considered valid if the number of ‘*’ and ‘#’ are equal. The ‘*’ and ‘#’ can be at any position in the string.
Note : The output will be a positive or negative integer based on number of ‘*’ and ‘#’ in the input string.

(*>#): positive integer
(#>*): negative integer
(#=*): 0
Example 1:
Input 1:

###***   -> Value of S
Output :

0   → number of * and # are equal

In [ ]:
string = input()

count_star = string.count("*")
count_hash = string.count("#")

if count_star == count_hash:
  print(0)
else:
  if count_star > count_hash:
    print(count_star - count_hash)
  else:
    print(count_hash - count_star)


hello*#*
1


In [ ]:
s=input()
a=0
b=0
for i in s:
    if i=='*':
        a+=1
    elif i=='#':
        b+=1
print(a-b)

***###
0


# A parking lot in a mall has RxC number of parking spaces. Each parking space will either be  empty(0) or full(1). The status (0/1) of a parking space is represented as the element of the matrix. The task is to find index of the prpeinzta row(R) in the parking lot that has the most of the parking spaces full(1).

Note :
RxC- Size of the matrix
Elements of the matrix M should be only 0 or 1.

Example 1:
Input :
3   -> Value of R(row)
3    -> value of C(column)
[0 1 0 1 1 0 1 1 1] -> Elements of the array M[R][C] where each element is separated by new line.
Output :
3  -> Row 3 has maximum number of 1’s



In [ ]:
r = int(input("Enter number of rows: "))
c = int(input("Enter number of columns: "))

max_sum = 0  # To track the maximum number of 1's in any row
max_row_index = 0  # To track the row index with the maximum number of 1's

for i in range(r):
    row_sum = 0  # Sum of 1's in the current row
    for j in range(c):
        value = int(input())  # Read the value at position (i, j)
        row_sum += value  # Accumulate the sum of 1's in the current row
    if row_sum > max_sum:
        max_sum = row_sum
        max_row_index = i + 1  # Row index is 1-based
    # No need to reset row_sum here as it is re-initialized at the start of each row

print(max_row_index)


Enter number of rows: 3
Enter number of columns: 3
1
0
1
1
1
0
1
1
1
3


In [ ]:
n = int(input())


for i in range(1,n+1):
  strng = input()
  print(len(strng))
print(" ")

3
one
3
two
3
three
5
 


In [ ]:
n = int(input())
lengths =[]

for i in range(n):
  word = input()
  length = len(word)
  lengths.append(length)

# Output the variable to STDOUT
for l in lengths:
    print(l)

3
six
seven
twenty two
3
5
10


In [ ]:
n = input()
numbers = list(map(int, n.split(',')))
add = sum(numbers)

product = 1

for i in numbers:
  product*=i

print("SUM = ", add)
print("PRODUCT =",product)

1,2,3,4,5
SUM =  15
PRODUCT = 120


In [ ]:
a = input()  # Input as a comma-separated string
a = a.split(",")  # Split the string into a list
a = list(map(int, a))  # Convert each element to an integer

largest = max(a)
smallest = min(a)

print("LARGEST =", largest)
print("SMALLEST =", smallest)

78,81,12,54,1010
LARGEST = 1010
SMALLEST = 12


# bQ13) You are given two arrays containing user IDs. array_1 has N user IDs and array_2 has all the user IDs i.e array_1 ID and an additional ID. Your task is to find out this additional ID.

Sample Input

1 4 7 3

1 2 3 4 5 6 7

Sample Output

2 5 6

Explanation

2, 5, and 6 are not present in the array_1 ID so that means its additional ID

In [ ]:
import numpy as np

lst1 = eval(input())
lst2 = eval(input())

set1 = set(lst1)
set2 = set(lst2)

result = set2.difference(set1)

new_lst1 = list(result)
print(*new_lst1)

[1,2,3]
[1,2,3,4,5,6]
4 5 6


In [ ]:
# Input arrays
array_1 = list(map(int, input().split()))  # Input for array_1
array_2 = list(map(int, input().split()))  # Input for array_2

# Convert both arrays to sets
set_1 = set(array_1)
set_2 = set(array_2)

# Find the additional IDs (elements in array_2 but not in array_1)
additional_ids = set_2 - set_1

# Print the result
print(*additional_ids)


1 4 7 3
1 2 3 4 5 6 7
2 5 6


# #1. Python Program for Check if all digits of a number divide it

# Input : 128
# Output : Yes
# 128 % 1 == 0, 128 % 2 == 0, and 128 % 8 == 0.

# Input : 130
# Output : No

In [ ]:
n = int(input())

num_str = str(n)  # Convert the number to a string to iterate over digits

for digit_char in num_str:
    digit = int(digit_char)

    if digit == 0 or n % digit != 0:  # Check if digit is 0 or if n is not divisible by the digit
        print("No")
        break
else:
    print("Yes")

2000
No


# #2. Given A and B, the task is to find the number of possible values
# that X can take such that the given modular equation (A mod X) = B holds good.
# Here, X is also called a solution of the modular equation.

# Input : A = 26, B = 2
# Output : 6
# Explanation
# X can be equal to any of {3, 4, 6, 8,
# 12, 24} as A modulus any of these values
# equals 2 i. e., (26 mod 3) = (26 mod 4)
# = (26 mod 6) = (26 mod 8) = Output: 2

# Input : 21 5
# Output : 2
# Explanation
# X can be equal to any of {8, 16} as A modulus
# any of these values equals 5 i.e. (21 mod
# 8) = (21 mod 16) = 5

In [ ]:
a = int(input())

b = int(input())

count = 0

for i in range(1,a+1):
  if a%i == b:
    count+=1
print(count)

21
5
2


# # 3. Given a number n, the task is to find the even factor sum of a number. Examples:
# Input : 30
# Output : 48
# Even dividers sum 2 + 6 + 10 + 30 = 48

# Input : 18
# Output : 26
# Even dividers sum 2 + 6 + 18 = 26


In [ ]:
n = int(input())

sum = 0

for i in range(1,n+1):
  if i % 2 == 0:
    if n % i == 0:
      sum+=i
print(sum)

30
48


# # 4. Given a string, write a python program to find the most occurrence character and its number of occurrences. Examples:

# Input : hello
# Output : ('l', 2)

# Input : geeksforgeeks
# Output : ('e', 4)

In [ ]:
string = input()
count = 0

lst = []

for i in range(len(string)):
  if i not in tup:
    lst.append(lst[i])
    tup = tuple(lst)
print(tup)


hello
(0, 1, 2, 3, 4)


In [ ]:
string = input()
lst = str(string)

freq = [lst.count(i) for i in lst]

my_dict = dict(zip(lst,freq))
print(my_dict)

hello siddharth how are you
{'h': 4, 'e': 2, 'l': 2, 'o': 3, ' ': 4, 's': 1, 'i': 1, 'd': 2, 'a': 2, 'r': 2, 't': 1, 'w': 1, 'y': 1, 'u': 1}


# # 5. Given three angles. The task is to check if it is possible to have a triangle of positive area with these angles.
# If it is possible print “YES” else print “NO”.

# Input : ang1 = 50, ang2 = 60, ang3 = 70
# Output : YES

# Input : ang1 = 50, ang2 = 65, ang3 = 80
# Output : NO

# Note:
# The sum of the three given angles equals to 180.
# The sum of any two angles is greater than equal to the third one.
# None of the given angles is zero.


In [ ]:
ang1 = int(input())
ang2 = int(input())
ang3 = int(input())

if ang1 + ang2 + ang3 == 180 and (ang1+ang2>=ang3 or ang1+ang3>=ang2 or ang2+ang3>=ang1):
  print("Yes")
else:
  print("No")

50
65
80
No


# # 6. Integers X and K are given. The task is to find smallest K-digit number divisible by X. Examples:

# Input : X = 83, K = 5
# Output : 10043
# 10043 is the smallest 5 digit
# number that is multiple of 83.

# Input : X = 5, K = 2
# Output : 10
# Hint: Get the smallest number of K digits (10**(k-1)) and then get the remainder divided by X, then subtract it to get the number.


In [ ]:
x = int(input())
k = int(input())

num = 10**(k-1)

while num % x !=0:
  num+=1
print(num)

5
2
10


# # 7. Given a square matrix of order n*n, we need to print elements of the matrix in Z form

# Examples:

# Input : mat=  [[1, 2, 3],
#                     [4, 5, 6],
#                     [7, 8, 9]]
# Output : 1 2 3 5 7 8 9
# Input : mat = [[5, 19, 8, 7],
#                    [4, 1, 14, 8],
#                    [2, 20, 1, 9],
#                    [1, 2, 55, 4]]
# Output: 5 19 8 7 14 20 1 2 55 4

In [ ]:
lst = eval(input())

for i in lst:
  print(*i)

[[5, 19, 8, 7], [4, 1, 14, 8], [2, 20, 1, 9], [1, 2, 55, 4]]
5 19 8 7
4 1 14 8
2 20 1 9
1 2 55 4


In [ ]:
lst = eval(input())
n = len(lst)

# Print the first row
for i in range(n):
    print(lst[0][i], end=" ")

# Print the diagonal (excluding first and last row)
for i in range(1, n-1):
    print(lst[i][n-i-1], end=" ")

# Print the last row (in reverse order)
for i in range(n):
    print(lst[-1][i], end=" ")

[[5, 19, 8, 7], [4, 1, 14, 8], [2, 20, 1, 9], [1, 2, 55, 4]]
5 19 8 7 14 20 1 2 55 4 

# # 8. Write a python program for a given long integer,
# we need to find if the difference between sum of odd position digits and sum of even position digits is 0 or not.
# Input: 1212112
# Output: Yes
# Explanation:
# the odd position element is 2+2+1=5
# the even position element is 1+1+1+2=5
# the difference is 5-5=0 equal to zero .
# So print yes.

# Input:12345
# Output: No
# Explanation:
# the odd position element is 1+3+5=9
# the even position element is 2+4=6
# the difference is 9-6=3 not equal to zero.
# So print no.

In [ ]:
n = int(input())

num = str(n)

sum_even = 0
sum_odd = 0
for i in range(len(num)):
  if i % 2 == 0:
    sum_even+=int(num[i])

  else:
    sum_odd+=int(num[i])

if sum_even - sum_odd == 0:
  print("Yes")
else:
  print("No")

12234
No


# # 9. Given a number, find minimum sum of its factors.
# Input : 12
# Output : 7
# Explanation:
# Following are different ways to factorize 12 and
# sum of factors in different ways.
# 12 = 12 * 1 = 12 + 1 = 13
# 12 = 2 * 6 = 2 + 6 = 8
# 12 = 3 * 4 = 3 + 4 = 7
# 12 = 2 * 2 * 3 = 2 + 2 + 3 = 7
# Therefore minimum sum is 7
# Input : 105
# Output : 15

In [ ]:
import math

def min_sum_of_factors(n):
    # Set initial min_sum as n+1 (maximum value possible sum)
    min_sum = n + 1

    # Loop through possible factors
    for i in range(1, int(math.sqrt(n)) + 1):
        if n % i == 0:  # Check if i is a factor
            # i and (n // i) are factor pairs
            factor1 = i
            factor2 = n // i
            # Sum the factors and check if it's the smallest
            min_sum = min(min_sum, factor1 + factor2)

    return min_sum

# Input
n = int(input("Enter a number: "))
print(min_sum_of_factors(n))

Enter a number: 12
7


# # 10. We have N coins which need to arrange in form of a triangle, i.e. first row will have 1 coin,
# second row will have 2 coins and so on,
# we need to tell maximum height which we can achieve by using these N coins. Examples:
# Input : N = 7
# Output : 3
# Explanation: Maximum height will be 3, putting 1, 2 and then 3 coins. It is not possible to use 1 coin left.

# Input : N = 12
# Output : 4
# Explanation: Maximum height will be 4, putting 1, 2, 3 and 4 coins, it is not possible to make height as 5, because that will require 15 coins.

In [ ]:
num = int(input())
count = 0
sum = 0

for i in range(1, num + 1):
    sum += i
    if sum > num:
        break
    count += 1

print(count)

12
4


# # 11. Given an integer ‘n’, write a Python function that returns true
# if binary representation of x is palindrome else return false. Examples:
# Input : n = 9
# Output : True
# Binary representation of n=9 is 1001 which
# is palindrome as well.

# Input : n = 10
# Output : False
# Binary representation of n=10 is 1010 which
# is not palindrome.

# 11. Given an integer ‘n’, write a Python function that returns true
# if binary representation of x is palindrome else return false. Examples:
# Input : n = 9
# Output : True
# Binary representation of n=9 is 1001 which
# is palindrome as well.

# Input : n = 10
# Output : False
# Binary representation of n=10 is 1010 which
# is not palindrome.

In [ ]:
n = int(input())

binary_str = bin(n)[2:]

if binary_str == binary_str[::-1]:
  print("True")
else:
  print("False")

9
True


In [ ]:
# Method 2
def decimal_to_binary(n):
    binary_digits = []

    # Convert decimal to binary by dividing by 2
    while n > 0:
        binary_digits.append(n % 2)  # Store remainder (either 0 or 1)
        n //= 2                      # Update n by dividing by 2

    # The binary digits are stored in reverse order, so reverse the list
    return binary_digits

def is_palindrome_in_binary(n):
    # Get the binary digits
    binary_digits = decimal_to_binary(n)

    # Check if the binary digits form a palindrome
    return binary_digits == binary_digits[::-1]

# Input
n = int(input("Enter a number: "))

# Check if the binary representation is a palindrome
if is_palindrome_in_binary(n):
    print(f"The binary representation of {n} is a palindrome.")
else:
    print(f"The binary representation of {n} is not a palindrome.")





Enter a number: 9
The binary representation of 9 is a palindrome.


# # 12. Given a value N, if we want to make change for N cents, and we have infinite supply of each of S = { S1, S2, .. , Sm} valued coins,
# how many ways can we make the change? The order of coins doesn\’t matter.
# For example, for N = 4 and S = {1,2,3}, there are four solutions: {1,1,1,1},{1,1,2},{2,2},{1,3}. So output should be 4.
# For N = 10 and S = {2, 5, 3, 6},
# there are five solutions: {2,2,2,2,2}, {2,2,3,3}, {2,2,6}, {2,3,5} and {5,5}. So the output should be 5.
#note: S is the valid coins  50 paise, one, two, five, ten, and twenty rupees for example : In India S ={ 0.5,1,2,5,10,20 }

4
{2,5,3,6}
0


# # 13. Python Program for Find largest prime factor of a number
# Input: 6
# Output: 3
# Explanation
# Prime factor of 6 are - 2, 3
# Largest of them is 3

# Input: 15
# Output: 5
 # Note: Input will be  > 2

6
4


# Convert the decimal into binary and add it’s bits

Sample input:
10

Output:
2

Explanation:
10 -> 1010 -> 1+0+1+0 -> 2

In [ ]:
def decimal_to_binary(n):
    binary_rep = ""  # Initialize an empty string to store binary digits
    while n > 0:
        binary_rep = str(n % 2) + binary_rep  # Prepend each bit to get the correct order
        n = n // 2
    return binary_rep

n = int(input("Enter a number: "))
binary_rep = decimal_to_binary(n)

sum_of_bits = 0

for bit in binary_rep:
    sum_of_bits += int(bit)

print("Binary representation:", binary_rep)
print("Sum of bits:", sum_of_bits)




Enter a number: 10
Binary representation: 1010
Sum of bits: 2


# Print the winning team of football match, the team with most goals is the winner

Sample input:
5

A
A
B
A
B

Output:
A

In [ ]:
n = int(input())

count_a = 0
count_b = 0

for i in range(n):
  team = input()
  if team == 'A':
    count_a+=1
  else:
    count_b+=1

if count_a > count_b:
  print("A")
else:
  print("B")

5
A
A
B
B
A
A


# # 14. Given two matrices X and Y, the task is to compute the sum of two matrices and then print it in Python.

# Input :
#  X= [[1,2,3],
#     [4 ,5,6],
#     [7 ,8,9]]

# Y = [[9,8,7],
#     [6,5,4],
#     [3,2,1]]

# Output :
#  result= [[10,10,10],
#          [10,10,10],
#          [10,10,10]]


In [ ]:
X = eval(input())
Y = eval(input())

# Initialize the result matrix with the same dimensions as X and Y
result = [[0 for j in range(len(X[0]))] for i in range(len(X))]

# Add corresponding elements of X and Y
for i in range(len(X)):
    for j in range(len(X[0])):
        result[i][j] = X[i][j] + Y[i][j]

# Print the entire result matrix
for row in result:
    print(row)

[[1,2,3], [4 ,5,6], [7 ,8,9]]
[[9,8,7], [6,5,4], [3,2,1]]
[10, 10, 10]
[10, 10, 10]
[10, 10, 10]


In [ ]:
# method 2
import numpy as np

X = np.array(eval(input()))
Y = np.array(eval(input()))

print(X+Y)

[[1,2,3], [4 ,5,6], [7 ,8,9]]
[[9,8,7], [6,5,4], [3,2,1]]
[[10 10 10]
 [10 10 10]
 [10 10 10]]
[[10 10 10]
 [10 10 10]
 [10 10 10]]


# # 15.  Python program to check whether the string is Symmetrical or Palindrome
# Input: khokho

# Output:
# The entered string is symmetrical
# The entered string is not palindrome

# Input:amaama
# Output:
# The entered string is symmetrical
# The entered string is palindrome

In [ ]:
string = input("Enter a string: ")

n = len(string)//2

# Split the string based on its length
if len(string) % 2 == 0:
    a = string[:n]
    b = string[n:]
else:
    a = string[:n]
    b = string[n+1:]

# Check symmetry and palindrome conditions
if a == b and string == string[::-1]:
    print("The entered string is symmetrical")
    print("The entered string is palindrome")
elif a == b:
    print("The entered string is symmetrical")
    print("The entered string is not palindrome")
elif string == string[::-1]:
    print("The entered string is not symmetrical")
    print("The entered string is palindrome")
else:
    print("The entered string is not symmetrical")
    print("The entered string is not palindrome")

Enter a string: nitin
The entered string is not symmetrical
The entered string is palindrome


# 16. Question:
# Write a program that calculates and prints the value according to the given formula:
# Q = Square root of [(2 * C * D)/H]
# Following are the fixed values of C and H:
# C is 50. H is 30.
# D is the variable whose values should be input to your program in a comma-separated sequence.
# Example
# Let us assume the following comma separated input sequence is given to the program:
# 100,150,180
# The output of the program should be:
# 18,22,24

# Hints:
# If the output received is in decimal form,
# it should be rounded off to its nearest value (for example, if the output received is 26.0, it should be printed as 26)
# In case of input data being supplied to the question, it should be assumed to be a console input.

In [ ]:
def calculate_values(input_sequence, C=50, H=30):
    # Split the input sequence into a list of integers
    values = [int(val) for val in input_sequence.split(',')]

    # Calculate the values using the formula
    result = [int(((2 * C * D) / H) ** 0.5) for D in values]

    return result

# Example input sequence
input_sequence = "100,150,180"

# Calculate and print the values
output_result = calculate_values(input_sequence)
print(','.join(map(str, output_result)))

18,22,24


# # 17. Write a program that accepts a comma separated sequence of words as input and prints the words in a comma-separated sequence after sorting them alphabetically.
# Suppose the following input is supplied to the program:
# without,hello,bag,world
# Then, the output should be:
# bag,hello,without,world

# Hints:
# In case of input data being supplied to the question, it should be assumed to be a console input. Try.sort()


In [ ]:
string = input()

string = string.split(',')

words = sorted(string)

print(','.join(words))

without,hello,bag,world
bag,hello,without,world


 # 18. Write a program which accepts a sequence of comma separated 4 digit binary numbers as its input
# and then check whether they are divisible by 5 or not. The numbers that are divisible by 5 are to be printed in a comma separated sequence.
# Example:
# Input : 0100,0011,1010,1001
# Then the output should be:
# 1010

In [ ]:
digit = input()

binary_digit = digit.split(',')

result = []
for i in binary_digit:
    if int(i, 2) % 5 == 0:
        result.append(i)

if result:
    print(','.join(result))


0100,0011,1010,1001,1111
1010,1111


# # 19. Write a program, which will find all such numbers between 1000 and 3000 (both included) such that each digit of the number is an even number.
# The numbers obtained should be printed in a comma-separated sequence on a single line.

In [ ]:
a = int(input())
b = int(input())

for i in range(a, b+1):
    if all(int(digit) % 2 == 0 for digit in str(i)):
        print(i, end=',')

1000
3000
2000,2002,2004,2006,2008,2020,2022,2024,2026,2028,2040,2042,2044,2046,2048,2060,2062,2064,2066,2068,2080,2082,2084,2086,2088,2200,2202,2204,2206,2208,2220,2222,2224,2226,2228,2240,2242,2244,2246,2248,2260,2262,2264,2266,2268,2280,2282,2284,2286,2288,2400,2402,2404,2406,2408,2420,2422,2424,2426,2428,2440,2442,2444,2446,2448,2460,2462,2464,2466,2468,2480,2482,2484,2486,2488,2600,2602,2604,2606,2608,2620,2622,2624,2626,2628,2640,2642,2644,2646,2648,2660,2662,2664,2666,2668,2680,2682,2684,2686,2688,2800,2802,2804,2806,2808,2820,2822,2824,2826,2828,2840,2842,2844,2846,2848,2860,2862,2864,2866,2868,2880,2882,2884,2886,2888,

# # 20. Write a program that accepts a sentence and calculate the number of letters and digits.
# Suppose the following input is supplied to the program:
# hello world! 123
# Then, the output should be:
# LETTERS 10
# DIGITS 3
# Hint : checkout : .isdigit() and .isalpha() method

In [ ]:
sentence = input()

letters_count = 0
digits_count = 0

for char in sentence:
  if char.isalpha():
    letters_count+=1
  elif char.isdigit():
    digits_count+=1

print("LETTERS", letters_count)
print("DIGITS", digits_count)


Hello World! 123
LETTERS 10
DIGITS 3


# # 21. Write a program that accepts a sentence and calculate the number of upper case letters and lower case letters.
# Suppose the following input is supplied to the program:
# Hello world!
# Then, the output should be:
# UPPER CASE 1
# LOWER CASE 9

# Hint : Checkout : .isupper() and .islower() method

In [ ]:
sentence = input()

upper_count = 0
lower_count = 0

for char in sentence:
  if char.isupper():
    upper_count+=1
  elif char.islower():
    lower_count+=1

print("UPPER CASE", upper_count)
print("LOWER CASE", lower_count)

Hello World 123
UPPER CASE 2
LOWER CASE 8


# 22. Write a program that computes the value of a+aa+aaa+aaaa (based on depth) with a given digit as the value of a.
# Suppose the following input is supplied to the program:
# depth = 4 and value of a = 9
# Then, the output should be:
# 11106


In [ ]:
depth = int(input())

value = int(input())

result = 0
current_term = 0
for i in range(1,depth+1):
  current_term=current_term*10+value
  result+=current_term

print(result)

4
9
11106


# 23. A robot moves in a plane starting from the original point (0,0). The robot can move toward UP, DOWN, LEFT and RIGHT with a given steps. The trace of robot movement is shown as the following:
# UP 5
# DOWN 3
# LEFT 3
# RIGHT 2
#
# The numbers after the direction are steps. Please write a program to compute the distance from current position after a sequence of movement and original point. If the distance is a float, then just print the nearest integer.
# Example:
# If the following tuples are given as input to the program:
# ((UP,5),(DOWN,3),(LEFT,3),(RIGHT,2))
# Then, the output of the program should be:
# 2

In [ ]:
import math
def Robot_move(movements):
  x,y = 0,0

  for direction,step in movements:
    if direction == 'UP':
      y+=step
    elif direction == 'DOWN':
      y-=step
    elif direction == 'LEFT':
      x-=step
    elif direction == 'RIGHT':
      x+=step

  return math.sqrt(x**2+y**2)

movements = [('UP',5),('DOWN',3),('LEFT',3),('RIGHT',2)]

distance = Robot_move(movements)

print(round(distance))



2


# # 24. Assuming that we have some email addresses in the "username@companyname.com" format, please write program to print the company name of a given email address. Both user names and company names are composed of letters only.

# Example:
# If the following email address is given as input to the program:

# john@google.com

# Then, the output of the program should be:

# google

In [ ]:
email = input()

email = email.split('@')

company_name = email[1].split('.')[0]

print(company_name)


john@google.com
google


# # 25. Write a Python program to
# find the sum of heights of the top three buildings in descending order from n given buildings.
# Input:
# n = 8
# Input the heights of eight buildings:
# 25
# 35
# 15
# 16
# 30
# 45
# 37
# 39
# output : 121

In [ ]:
n = int(input())

heights = []

for i in range(n):
  height_inputs = int(input())
  heights.append(height_inputs)

total_height = (sorted(heights))
top_three_height = sum(total_height[-3:])
print(top_three_height)

8
25
35
15
16
30
45
37
39
121
121


# # 26. Write a Python program to check whether three given lengths (integers) of three sides form a right triangle. Print "Yes" if the given sides form a right triangle otherwise print "No".
# Input:
# Integers separated by a single space.
# Input three integers(sides of a triangle)
# 8 6 7
# No

In [ ]:
sides = input()
sides = sides.split(' ')

lst_sides = list(map(int,sides))

a,b,c = sorted(lst_sides)

if a**2 + b**2 == c**2:
  print("Yes")
else:
  print("No")

6 7 8
No


## 27. There are two circles C1 with radius r1, central coordinate (x1, y1) and C2 with radius r2 and central coordinate (x2, y2).

# Write a Python program to test the followings -

# "C2 is in C1" if C2 is in C1
# "C1 is in C2" if C1 is in C2
# "Circumference of C1 and C2 intersect" if circumference of C1 and C2 intersect
# "C1 and C2 do not overlap" if C1 and C2 do not overlap and
# "Circumference of C1 and C2 will touch" if C1 and C2 touch

# Hint :
# Let the distance between the two centers be  𝑑
#  , and the radiuses be  𝑟1,𝑟2
#   with  𝑟1≤𝑟2

# If you have  𝑑<𝑟2−𝑟1
#  , then they don’t touch, and the smaller circle is inside the larger circle.

# If you have  𝑑=𝑟2−𝑟1
#  , then they touch at one point internally.

# If you have  𝑟2−𝑟1<𝑑<𝑟2+𝑟1
#  , then they touch at two points.

# If you have  𝑑=𝑟2+𝑟1
#  , then they touch at one point externally.

# If you have  𝑑>𝑟2+𝑟1
#  , then they don’t touch, and neither is inside the other.

# If you know that it is either touching internally or externally — that is, by the chart above, that  𝑑=𝑟2−𝑟1
#   or  𝑑=𝑟2+𝑟1
#  , then you can simplify things.

# If you have  𝑑2>𝑟22,
#   then they touch externally.

# If you have  𝑑2<𝑟22
#  , then they touch internally.

# Computing  𝑑 = sqaure root of ((𝑥2−𝑥1)^2+(𝑦2−𝑦1)^2)
#   involves a square root calculation, which is why I used  𝑑2
#   for the “simplified” version.

# # 29. If you draw a straight line on a plane, the plane is divided into two regions.
# For example, if you draw two straight lines in parallel, you get three areas, and if you draw vertically one to the other you get 4 areas.
# Write a Python program to create the maximum number of regions obtained by drawing n given straight lines.
# Input:
# (1 <= n <= 10,000)
# Input number of straight lines (o to exit):
# 5
# Number of regions:
# 16
# Hint : Formula to calculate number of regions  (n*n+n+2)//2

In [ ]:
n = int(input())

regions = (n*n+n+2)//2

print("Number of regions:",regions)

5
Number of regions: 16


# # 31. Write a Python program to find the longest common prefix string among a given array of strings. Return false if there is no common prefix.
# For Example, longest common prefix of "abcdefgh" and "abcefgh" is "abc".
# Sample Input:
# ["abcdefgh","abcefgh"]
# ["w3r","w3resource"]
# ["Python","PHP", "Perl"]
# ["Python","PHP", "Java"]
# Sample Output:
# abc
# w3r
# P

In [ ]:
def longest_common_prefix(strings):
    if not strings:
        return False

    # Sort the strings to minimize the comparisons
    strings.sort()

    # The common prefix must be in the shortest string
    prefix = strings[0]

    # Compare the shortest string with the rest of the strings
    for i in range(1, len(strings)):
        while strings[i].find(prefix) != 0:
            # Shorten the prefix until a common one is found
            prefix = prefix[:-1]
            if not prefix:
                return False  # Return False if no common prefix found

    return prefix

# Test cases
print(longest_common_prefix(["abcdefgh", "abcefgh"]))  # Output: "abc"
print(longest_common_prefix(["w3r", "w3resource"]))    # Output: "w3r"
print(longest_common_prefix(["Python", "PHP", "Perl"]))  # Output: "P"
print(longest_common_prefix(["Python", "PHP", "Java"]))  # Output: False


abc
w3r
P
False


# # 32. The price of a given stock on each day is stored in an array.
# Write a Python program to find the maximum profit in one transaction i.e., buy one and sell one share of the stock from the given price value of the said array. You cannot sell a stock before you buy one.
# Input (Stock price of each day): [224, 236, 247, 258, 259, 225]
# Output: 35
# Explanation:
# 236 - 224 = 12
# 247 - 224 = 23
# 258 - 224 = 34
# 259 - 224 = 35
# 225 - 224 = 1
# 247 - 236 = 11
# 258 - 236 = 22
# 259 - 236 = 23
# 225 - 236 = -11
# 258 - 247 = 11
# 259 - 247 = 12
# 225 - 247 = -22
# 259 - 258 = 1
# 225 - 258 = -33
# 225 - 259 = -34

In [ ]:
lst = eval(input())

max = 0

for i in range(len(lst)):

  for j in range(i+1,len(lst)):

    result = lst[j]-lst[i]

    if result > max:
      max = result

print(max)

[224, 236, 247, 258, 259, 225]
35


# # 33. Write a Python program to compute the largest product of three integers from a given list of integers.
# Sample Input:
# [-10, -20, 20, 1]
# [-1, -1, 4, 2, 1]
# [1, 2, 3, 4, 5, 6]
# Sample Output:
# 4000
# 8
# 120


In [ ]:
lst = eval(input())

lst.sort()

result  = lst[-1]*lst[-2]*lst[-3]

print(result)

[-10, -20, 20, 1]
-200


# # 34. Write a Python program that accepts a list of numbers. Count the negative numbers and compute the sum of the positive numbers of the said list. Return these values through a list.
# Original list: [1, 2, 3, 4, 5]
# Number of negative of numbers and sum of the positive numbers of the said list: [0, 15]
# Original list: [-1, -2, -3, -4, -5]
# [5, 0]
# Number of negative of numbers and sum of the positive numbers of the said list: [5, 0]
# Original list: [1, 2, 3, -4, -5]
# [2, 6]
# Number of negative of numbers and sum of the positive numbers of the said list: [2, 6]
# Original list: [1, 2, -3, -4, -5]
# [3, 3]
# Number of negative of numbers and sum of the positive numbers of the said list: [3, 3]

In [ ]:
original_list = eval(input())

negative_numbers = 0

sum_positive_numbers = 0

for i in original_list:
  if i < 0:
    negative_numbers+=1
  else:
    sum_positive_numbers+=i

result = [negative_numbers, sum_positive_numbers]

print(result)

[-1, -2, -3, -4, -5]
[5, 0]


# # 35. Write a Python program to check whether a sequence of numbers has an increasing trend or not.
# Sample Input:
# [1,2,3,4]
# [1,2,5,3,4]
# [-1,-2,-3,-4]
# [-4,-3,-2,-1]
# [1,2,3,4,0]
# Sample Output:
# True
# False
# False
# True
# False

In [ ]:
lst = eval(input())
increasing_trend = True

for i in range(len(lst)-1):
  if lst[i] >= lst[i+1]:
    increasing_trend = False
    break

print(increasing_trend)

[[-1,-2,-3,-4]]
True


#  30. When character are consecutive in a string , it is possible to shorten the character string by replacing the character with a certain rule.
# For example, in the case of the character string YYYYY, if it is expressed as @ 5 Y, it is compressed by one character.
# Write a Python program to restore the original string by entering the compressed string with this rule. However, the @ character does not appear in the restored character string.
# Input:
# The restored character string for each character on one line.
# Original text: XY@6Z1@4023
# XYZZZZZZ1000023
# Original text: @39+1=1@30
# 999+1=1000
# Consider 1-9 after @

In [ ]:
input_text = input()

output_text = ''

i = 0

while i<len(input_text):
  if input_text[i]!='@':
    output_text+=input_text[i]
    i = i+1

  else:
    output_text+=input_text[i+2]*int(input_text[i+1])
    input_text = input_text[i+3:]
    i = 0

print(output_text)

@39+1=1@30
999+1=1000


# # 36. Write a Python program to check whether two given lines are parallel or not.
# Note: Parallel lines are two or more lines that never intersect. Parallel Lines are like railroad tracks that never intersect.
# The General Form of the equation of a straight line is: ax + by = c
# The said straight line is represented in a list as [a, b, c]
# Example of two parallel lines:
# x + 4y = 10 and x + 4y = 14
# Sample Input:
# ([2,3,4], [2,3,8])
# ([2,3,4], [4,-3,8])
# Sample Output:
# True
# False

In [ ]:
# For condition pf parallel lines a1.b2 = a2.b1
def are_lines_parallel(line1, line2):
    # Check if the coefficients are proportional
    return line1[0] * line2[1] == line2[0] * line1[1]

# Sample input
input_lines = [
    ([2, 3, 4], [2, 3, 8]),
    ([2, 3, 4], [4, -3, 8])
]

# Sample output
for lines in input_lines:
    result = are_lines_parallel(*lines)
    print(result)

True
False


# # 37. Write a Python program to reverse all words of odd lengths.
# Sample Input:
# ("The quick brown fox jumps over the lazy dog")
# ("Python Exercises")
# Sample Output:
# The quick brown fox jumps revo the yzal dog
# nohtyP Exercises


In [ ]:
string = input()
string = string.split()

strng = ''
for i in range(len(string)):
  if i%2==0:
    strng+=string[i]
  else:
    strng+=string[i][::-1]

print(strng)

python exercide
pythonedicrexe


In [ ]:
def reverse_odd_length_words(sentence):
    words = sentence.split()
    reversed_words = [word[::-1] if len(word) % 2 == 0 else word for word in words]
    reversed_sentence = ' '.join(reversed_words)
    return reversed_sentence

# Sample input
input_sentences = [
    "The quick brown fox jumps over the lazy dog",
    "Python Exercises"
]

# Sample output
for sentence in input_sentences:
    result = reverse_odd_length_words(sentence)
    print(result)

The quick brown fox jumps revo the yzal dog
nohtyP Exercises


# 1
# 2    1
# 4    2    1
# 8    4    2    1
# 16    8    4    2    1
# 32   16    8    4    2    1
# 64   32   16    8    4    2    1
# 128   64   32   16    8    4    2    1

# Print the above pattern

In [ ]:
n = int(input())

for i in range(1,n+1):
  for j in range(i-1,-1,-1):
    print(2**(j), end = " ")
  print()


8
1 
2 1 
4 2 1 
8 4 2 1 
16 8 4 2 1 
32 16 8 4 2 1 
64 32 16 8 4 2 1 
128 64 32 16 8 4 2 1 


# 39. Create a function that takes a string s and returns a list of two-paired characters. If the string has an odd number of characters, add an asterisk * in the final pair.

# See the below examples for a better understanding:

# Examples
# string_pairs("mubashir") ➞ ["mu", "ba", "sh", "ir"]

# string_pairs("edabit") ➞ ["ed", "ab", "it"]

# string_pairs("airforces") ➞ ["ai", "rf", "or", "ce", "s*"]
# Notes
# Return [] if the given string is empty.

In [ ]:
string = input()
result = []

for i in range(0,len(string),2):
  if i+1 < len(string):
    result.append(string[i:i+2])
  else:
    result.append(string[i]+'*')
print(result)

edibles
['ed', 'ib', 'le', 's*']


# 40. Write a function that takes a credit card number and only displays the last four characters. The rest of the card number must be replaced by ************.

# Examples
# card_hide("1234123456785678") ➞ "************5678"

# card_hide("8754456321113213") ➞ "************3213"

# card_hide("35123413355523") ➞ "**********5523"
# Examples
# Ensure you return a string.
# The length of the string must remain the same as the input.

In [ ]:
number = int(input())

result = '*'*(len(str(number))-4)+str(number)[-4:]

print(result)

1234567988881111
************1111


# 41. Transcribe the given DNA strand into corresponding mRNA - a type of RNA, that will be formed from it after transcription. DNA has the bases A, T, G and C, while RNA converts to U, A, C and G respectively.

# Examples
# dna_to_rna("ATTAGCGCGATATACGCGTAC") ➞ "UAAUCGCGCUAUAUGCGCAUG"

# dna_to_rna("CGATATA") ➞ "GCUAUAU"

# dna_to_rna("GTCATACGACGTA") ➞ "CAGUAUGCUGCAU"
# Notes
# Transcription is the process of making complementary strand.
# A, T, G and C in DNA converts to U, A, C and G respectively, when in mRNA.

In [ ]:
def dna_to_rna(string):
  dictionary = {'A':'U','T':'A','G':'C','C':'G'}
  result = ''
  for i in string:
    result+=dictionary[i]
  return result
string = input()
print(dna_to_rna(string))

CGATATA
GCUAUAU


"Loves me, loves me not" is a traditional game in which a person plucks off all the petals of a flower one by one, saying the phrase "Loves me" and "Loves me not" when determining whether the one that they love, loves them back.

# Given a number of petals, return a string which repeats the phrases "Loves me" and "Loves me not" for every alternating petal, and return the last phrase in all caps. Remember to put a comma and space between phrases.

# Examples
# loves_me(3) ➞ "Loves me, Loves me not, LOVES ME"

# loves_me(6) ➞ "Loves me, Loves me not, Loves me, Loves me not, Loves me, LOVES ME NOT"

# loves_me(1) ➞ "LOVES ME"
# Notes
# Remember to return a string.
# The first phrase is always "Loves me".

In [ ]:
def loves_me(n):
    phrases = []

    # Loop through each petal
    for i in range(0, n):
        if i % 2 == 0:
            phrases.append("Loves me")
        else:
            phrases.append("Loves me not")

    # Capitalize the last phrase
    phrases[-1] = phrases[-1].upper()

    # Join the phrases with ', ' and return as a single string
    return ', '.join(phrases)

# Test cases
n = int(input())
print(loves_me(n))


6
Loves me, Loves me not, Loves me, Loves me not, Loves me, LOVES ME NOT


# Consider the following Python dictionary data and Python list labels:

data = {'birds': ['Cranes', 'Cranes', 'plovers', 'spoonbills', 'spoonbills', 'Cranes', 'plovers', 'Cranes', 'spoonbills', 'spoonbills'], 'age': [3.5, 4, 1.5, np.nan, 6, 3, 5.5, np.nan, 8, 4], 'visits': [2, 4, 3, 4, 3, 4, 2, 2, 3, 2], 'priority': ['yes', 'yes', 'no', 'yes', 'no', 'no', 'no', 'yes', 'no', 'no']}

labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

In [ ]:
#1. Create a DataFrame birds from this dictionary data which has the index labels.
import numpy as np
import pandas as pd

data = {'birds': ['Cranes', 'Cranes', 'plovers', 'spoonbills', 'spoonbills', 'Cranes', 'plovers', 'Cranes', 'spoonbills', 'spoonbills'], 'age': [3.5, 4, 1.5, np.nan, 6, 3, 5.5, np.nan, 8, 4], 'visits': [2, 4, 3, 4, 3, 4, 2, 2, 3, 2], 'priority': ['yes', 'yes', 'no', 'yes', 'no', 'no', 'no', 'yes', 'no', 'no']}

labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

birds = pd.DataFrame(data, index = labels)
birds

birds  age  visits priority
a      Cranes  3.5       2      yes
b      Cranes  4.0       4      yes
c     plovers  1.5       3       no
d  spoonbills  NaN       4      yes
e  spoonbills  6.0       3       no
f      Cranes  3.0       4       no
g     plovers  5.5       2       no
h      Cranes  NaN       2      yes
i  spoonbills  8.0       3       no
j  spoonbills  4.0       2       no

In [ ]:
#2. Display a summary of the basic information about birds DataFrame and its data.
birds.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10 entries, a to j
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   birds     10 non-null     object 
 1   age       8 non-null      float64
 2   visits    10 non-null     int64  
 3   priority  10 non-null     object 
dtypes: float64(1), int64(1), object(2)
memory usage: 700.0+ bytes


In [ ]:
birds.describe()

age     visits
count  8.000000  10.000000
mean   4.437500   2.900000
std    2.007797   0.875595
min    1.500000   2.000000
25%    3.375000   2.000000
50%    4.000000   3.000000
75%    5.625000   3.750000
max    8.000000   4.000000

In [ ]:
#3. Print the first 2 rows of the birds dataframe
birds.head(2)

birds  age  visits priority
a  Cranes  3.5       2      yes
b  Cranes  4.0       4      yes

In [ ]:
#4. Print all the rows with only 'birds' and 'age' columns from the dataframe
birds[['birds','age']]

birds  age
a      Cranes  3.5
b      Cranes  4.0
c     plovers  1.5
d  spoonbills  NaN
e  spoonbills  6.0
f      Cranes  3.0
g     plovers  5.5
h      Cranes  NaN
i  spoonbills  8.0
j  spoonbills  4.0

In [ ]:
#5. select [2, 3, 7] rows and in columns ['birds', 'age', 'visits']

birds.iloc[[2,3,7]][['birds','age','visits']]

birds  age  visits
c     plovers  1.5       3
d  spoonbills  NaN       4
h      Cranes  NaN       2

In [ ]:
#6. select the rows where the number of visits is less than 4
birds[birds['visits']<4]

birds  age  visits priority
a      Cranes  3.5       2      yes
c     plovers  1.5       3       no
e  spoonbills  6.0       3       no
g     plovers  5.5       2       no
h      Cranes  NaN       2      yes
i  spoonbills  8.0       3       no
j  spoonbills  4.0       2       no

In [ ]:
#7. select the rows with columns ['birds', 'visits'] where the age is missing i.e NaN
birds[birds['age'].isnull()][['birds','visits']]

birds  visits
d  spoonbills       4
h  trumpeters       2

In [ ]:
#8. Select the rows where the birds is a Cranes and the age is less than 4
birds[(birds['birds'] == 'Cranes') & (birds['age'] < 4)]

birds  age  visits priority
a  Cranes  3.5       2      yes
f  Cranes  3.0       4       no

In [ ]:
#9. Select the rows the age is between 2 and 4(inclusive)

birds[(birds['age'] >= 2) & (birds['age'] <= 4)][['birds', 'age', 'visits', 'priority']]

birds  age  visits priority
a      Cranes  3.5       2      yes
b      Cranes  4.0       4      yes
f      Cranes  3.0       4       no
j  spoonbills  4.0       2       no

In [ ]:
#10. Find the total number of visits of the bird Cranes
birds[birds['birds']=='Cranes']['visits'].sum()

12

In [ ]:
#11. Calculate the mean age for each different birds in dataframe.
birds.groupby('birds')['age'].mean()

birds
Cranes        3.5
plovers       3.5
spoonbills    6.0
Name: age, dtype: float64

In [ ]:
#12. Append a new row 'k' to dataframe with your choice of values for each column. Then delete that row to return the original DataFrame.
birds.loc['k'] = ['parrot', 5, 3, 'yes']
birds

birds  age  visits priority
a      Cranes  3.5       2      yes
b      Cranes  4.0       4      yes
c     plovers  1.5       3       no
d  spoonbills  NaN       4      yes
e  spoonbills  6.0       3       no
f      Cranes  3.0       4       no
g     plovers  5.5       2       no
h      Cranes  NaN       2      yes
i  spoonbills  8.0       3       no
j  spoonbills  4.0       2       no
k      parrot  5.0       3      yes

In [ ]:
birds = birds.drop(index='k')
birds

birds  age  visits priority
a      Cranes  3.5       2      yes
b      Cranes  4.0       4      yes
c     plovers  1.5       3       no
d  spoonbills  NaN       4      yes
e  spoonbills  6.0       3       no
f      Cranes  3.0       4       no
g     plovers  5.5       2       no
h      Cranes  NaN       2      yes
i  spoonbills  8.0       3       no
j  spoonbills  4.0       2       no

In [ ]:
#13. Find the number of each type of birds in dataframe (Counts)
birds['birds'].value_counts()

birds
Cranes        4
spoonbills    4
plovers       2
Name: count, dtype: int64

In [ ]:
#14. Sort dataframe (birds) first by the values in the 'age' in decending order, then by the value in the 'visits' column in ascending order.
birds.sort_values(by=['age', 'visits'], ascending=[False, True])

birds  age  visits priority
i  spoonbills  8.0       3       no
e  spoonbills  6.0       3       no
g     plovers  5.5       2       no
j  spoonbills  4.0       2       no
b      Cranes  4.0       4      yes
a      Cranes  3.5       2      yes
f      Cranes  3.0       4       no
c     plovers  1.5       3       no
h      Cranes  NaN       2      yes
d  spoonbills  NaN       4      yes

In [ ]:
#16. In the 'birds' column, change the 'Cranes' entries to 'trumpeters'.
birds['birds'] = birds['birds'].replace('Cranes', 'trumpeters')
birds

birds  age  visits priority
a  trumpeters  3.5       2      yes
b  trumpeters  4.0       4      yes
c     plovers  1.5       3       no
d  spoonbills  NaN       4      yes
e  spoonbills  6.0       3       no
f  trumpeters  3.0       4       no
g     plovers  5.5       2       no
h  trumpeters  NaN       2      yes
i  spoonbills  8.0       3       no
j  spoonbills  4.0       2       no

In [ ]:
#15. Replace the priority column values with'yes' should be 1 and 'no' should be 0
birds['priority'] = birds['priority'].replace({'yes': 1, 'no': 0})
birds

birds  age  visits  priority
a  trumpeters  3.5       2         1
b  trumpeters  4.0       4         1
c     plovers  1.5       3         0
d  spoonbills  NaN       4         1
e  spoonbills  6.0       3         0
f  trumpeters  3.0       4         0
g     plovers  5.5       2         0
h  trumpeters  NaN       2         1
i  spoonbills  8.0       3         0
j  spoonbills  4.0       2         0